### ISSUE TITLE
module outputs eaten by TF in 0.8.7

### ISSUE THREAD

**Issue by eyespies** [hashicorp/terraform#12240]
Hi there,

### Terraform Version
Broken in 0.8.7, works fine in 0.8.6

### Affected Resource(s)
Any module outputs

### Terraform Configuration Files
I'm open to discussing how to get this to you as our setup involves about 30 modules and a large set of resources.

### Debug Output
I couldn't create a Gist - it seems I crashed Github when I tried (I got a 405 error). Here are S3 links:
https://s3-us-west-2.amazonaws.com/bqdevops-public/terraform/tf-0.8.6-apply.txt (works as expected)
https://s3-us-west-2.amazonaws.com/bqdevops-public/terraform/tf-0.8.6-plan.txt (works as expected)
https://s3-us-west-2.amazonaws.com/bqdevops-public/terraform/tf-0.8.7-apply.txt (deletes my module outputs)
https://s3-us-west-2.amazonaws.com/bqdevops-public/terraform/tf-0.8.7-plan.txt (deletes my module outputs)

### Panic Output
None

### Expected Behavior
When I run a terraform apply in one project (ops-vpc) and export the remote state, I should see output like the following:
```
BQs-MacBook-Pro:ops-us-west-2 justin.spies$ ta vpc.plan 
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

chef-handler-sns-arn = arn:aws:sns:us-west-2:xxxxxxxxxxx:chef-pprod
ops-az = us-west-2a
ops-az-1 = us-west-2a
ops-az-2 = us-west-2b
ops-az-3 = us-west-2c
ops-db-subnet-1-cidr = 10.2.4.0/24
ops-db-subnet-1-id = subnet-xxxxx
ops-db-subnet-2-cidr = 10.2.5.0/24
ops-db-subnet-2-id = subnet-xxxxx
ops-db-subnet-3-cidr = 10.2.8.0/24
ops-db-subnet-3-id = subnet-xxxxx
ops-private-route-1-id = rtb-xxxxx
ops-private-route-2-id = rtb-xxxxx
ops-private-route-3-id = rtb-xxxxx
ops-private-subnet-1-cidr = 10.2.2.0/24
ops-private-subnet-1-id = subnet-xxxxx
ops-private-subnet-2-cidr = 10.2.3.0/24
ops-private-subnet-2-id = subnet-xxxxx
ops-private-subnet-3-cidr = 10.2.7.0/24
ops-private-subnet-3-id = subnet-xxxxx
ops-private-subnet-acl-id = acl-xxxxx
ops-public-route-1-id = rtb-xxxxx
ops-public-route-2-id = rtb-xxxxx
ops-public-route-3-id = rtb-xxxxx
ops-public-subnet-1-cidr = 10.2.0.0/24
ops-public-subnet-1-id = subnet-xxxxx
ops-public-subnet-2-cidr = 10.2.1.0/24
ops-public-subnet-2-id = subnet-xxxxx
ops-public-subnet-3-cidr = 10.2.6.0/24
ops-public-subnet-3-id = subnet-xxxxx
ops-vpc-cidr = 10.2.0.0/16
ops-vpc-id = vpc-xxxxx
ops-vpc-igw = igw-xxxxx
BQs-MacBook-Pro:ops-us-west-2 justin.spies$
```

Additionally, I would expect to see those outputs in my terraform.tfstate file, e.g.:
```
{
            "path": [
                "root",
                "bq-ops-vpc"
            ],
            "outputs": {
                "ops-db-subnet-1-id": {
                    "sensitive": false,
                    "type": "string",
                    "value": "subnet-xxxxx"
                },
                "ops-db-subnet-2-id": {
                    "sensitive": false,
                    "type": "string",
                    "value": "subnet-xxxxx"
                },
```

### Actual Behavior
In TF 0.8.6, this works fine. In TF 0.8.7 however, I see the following output after the apply:
```
BQs-MacBook-Pro:ops-us-west-2 justin.spies$ ta vpc.plan 

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

ops-az = us-west-2a
ops-az-1 = us-west-2a
ops-az-2 = us-west-2b
ops-az-3 = us-west-2c
ops-db-subnet-1-cidr = 10.2.4.0/24
ops-db-subnet-2-cidr = 10.2.5.0/24
ops-db-subnet-3-cidr = 10.2.8.0/24
ops-private-subnet-1-cidr = 10.2.2.0/24
ops-private-subnet-2-cidr = 10.2.3.0/24
ops-private-subnet-3-cidr = 10.2.7.0/24
ops-public-subnet-1-cidr = 10.2.0.0/24
ops-public-subnet-2-cidr = 10.2.1.0/24
ops-public-subnet-3-cidr = 10.2.6.0/24
ops-vpc-cidr = 10.2.0.0/16
BQs-MacBook-Pro:ops-us-west-2 justin.spies$ 
```

And I see the following contents in my state file (note that ALL the outputs are missing):
```
{
            "path": [
                "root",
                "bq-ops-vpc"
            ],
            "outputs": {},
```

### Steps to Reproduce
This may be specific to my environment, but I can't imagine why, and since it works in TF 0.8.6, it seems pretty likely that a code change broke this.
1. `tp -target=module.bq-ops-vpc  -out=vpc.plan -Xlegacy-graph`
1. `ta vpc.plan`

I used aliases in the above, here they are:
```
BQs-MacBook-Pro:ops-us-west-2 justin.spies$ alias | grep terraf
alias ta='terraform apply '
alias tg='terraform get -update=true'
alias to='terraform output '
alias tp='terraform plan -var-file=variables/aws.tfvars -var-file=variables/env.tfvars '
alias ts='terraform show '
```

### Important Factoids
All AWS stuff.

### References
None

**Comment by Nomon** [comment]
I ran into the same issue. I created a plan file by targeting a module directly under the root (module.module_name), applied it and the resulting state was missing all but one output. The one remaining was used as an input to the target module.

**Comment by leanderjanssen** [comment]
I also noticed that the output is still there if its value is defined as a variable, e.g.
```
output "role-concourse-worker-region" {
  value = "${var.region}"
}
```

When the output is defined with a value that comes from a resource attribute, the output dissapears from the state:

```
output "role-arn-concourse-worker" {
  value = "${aws_iam_role.role-concourse-worker.arn}"
}
```

**Comment by vtorhonen** [comment]
Experiencing the same issue with Terraform version 0.8.8. A workaround for this is to run `terraform refresh` after applying a plan.

**Comment by apparentlymart** [comment]
I expect this was caused by #11921, which was in 0.8.7. I believe the intent of that change was that we wouldn't try to _update_ non-targeted outputs, but it seems like the effect was that they now don't get included at all.

**Comment by Pryz** [comment]
@apparentlymart https://github.com/hashicorp/terraform/pull/11921 doesn't seem to be in 0.8.7 : 

```
$ git tag --contains 4d6085b46ac5e59c383303c9eb4d3d25ded2967e
v0.9.0
v0.9.0-beta1
v0.9.0-beta2
v0.9.1
v0.9.2
v0.9.3
v0.9.4
```

**Comment by apparentlymart** [comment]
Hmm... you're right. There is an entry in the changelog for 0.8.7 that had an incorrect PR reference, so I did some detective work to try to figure out what PR it actually was, but indeed it seems like I got the wrong one. I'll have another look for it tomorrow.

In the mean time, I've been having trouble reproducing this issue on master and now I'm wondering if the graph builder reworks in 0.9 have fixed this issue. Has anyone been able to see this error on a 0.9 release?

**Comment by whiteadam** [comment]
I am no longer seeing this bug myself. @eyespies, can you confirm?

**Comment by apparentlymart** [comment]
In the next release we will have #14186, which may end up addressing this issue in an _indirect_ way: it'll ensure that if you `-target` something that is used in an output, the output is also implicitly targeted. In the current stable version this isn't true, and so `-target` ends up preventing root outputs from getting updated by excluding them from the apply graph.

**Comment by eyespies** [comment]
Since upgrading to 0.9.4, this has not been an issue.

**Comment by apparentlymart** [comment]
Thanks for confirming, @eyespies!

While it's not really satisfying to just see the issue go away without being able to explain what fixed it, we changed a lot about how the graphs are built since 0.8.7 so it's reasonable to assume that this got fixed as a side-effect of some of that code re-organization.

So I'm going to close this for now. If anyone sees a similar issue on Terraform 0.9 or above, it'd be helpful to have it filed as a new top-level issue, since things are different enough in 0.9 that we'd likely need fresh debug output, repro steps, etc in order to really get to the bottom of it.

Thanks for all great discussion and information-sharing here, everyone!
